File: 621-task-scheduler/621-task-scheduler.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        hashMap = dict()
        
        for task in tasks:
            hashMap[task] = hashMap.get(task, 0) + 1
        
        sortedItems = sorted(hashMap.values())
        
        maxFreq = sortedItems[-1]
        idleTime = (maxFreq - 1) * n
        
        for i in range(len(sortedItems)-2, -1, -1):
            idleTime -= min(sortedItems[i], maxFreq - 1)
            
        idleTime = max(0, idleTime)
        
        return len(tasks) + idleTime
        

File: 621-task-scheduler/621-task-scheduler.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        return solve(tasks, n);
    }
private:
    // Rahul Sir
    int solve(vector<char>& tasks, int n) {
        unordered_map<char, int> hashMap;
        
        for(auto task: tasks)
            hashMap[task]++;
        vector<int> values;
        for(auto x: hashMap)
            values.push_back(x.second);
        
        sort(values.begin(), values.end());
        
        int maxFreq = values.back();
        int idleTime = (maxFreq - 1) * n;
        
        for(int i = values.size() - 2; i >= 0; --i) {
            idleTime -= min(values[i], maxFreq - 1);
        }
        idleTime = idleTime < 0 ? 0: idleTime;
        
        return tasks.size() + idleTime;
    }
};